contracts/contracts/Staking.sol

```diff
@@ -0,0 +1,225 @@
+// SPDX-License-Identifier: Apache-2.0
+
+pragma solidity ^0.8.0;
+
+import {Subcall, StakingAddress} from "./Subcall.sol";
+
+/**
+ * @title Minimal Staking Implementation
+ *
+ * This contract implements delegation and undelegation of the native ROSE token
+ * with validators. It encompasses the Oasis specific staking logic.
+ */
+contract Staking {
+    /// Incremented counter to determine receipt IDs
+    uint64 private lastReceiptId;
+
+    mapping(uint64 => PendingDelegation) private pendingDelegations;
+
+    /// (from, to) => shares
+    mapping(address => mapping(StakingAddress => uint128)) private delegations;
+
+    /// (receiptId => PendingUndelegation)
+    mapping(uint64 => PendingUndelegation) private pendingUndelegations;
+
+    /// (endReceiptId => UndelegationPool)
+    mapping(uint64 => UndelegationPool) private undelegationPools;
+
+    // -------------------------------------------------------------------------
+
+    struct PendingDelegation {
+        address from;
+        StakingAddress to;
+        uint128 amount;
+    }
+
+    struct PendingUndelegation {
+        StakingAddress from;
+        address payable to;
+        uint128 shares;
+        uint64 endReceiptId;
+        uint64 epoch;
+    }
+
+    struct UndelegationPool {
+        uint128 totalShares;
+        uint128 totalAmount;
+    }
+
+    // -------------------------------------------------------------------------
+
+    event OnDelegateStart(
+        address indexed who,
+        StakingAddress to,
+        uint256 amount,
+        uint64 receiptId
+    );
+
+    event OnDelegateDone(uint64 indexed receiptId, address who, uint128 shares);
+
+    event OnUndelegateStart(
+        uint64 indexed receiptId,
+        address who,
+        uint64 epoch,
+        uint128 shares
+    );
+
+    // -------------------------------------------------------------------------
+
+    constructor() {
+        // Due to an oddity in the oasis-cbor package, we start at 2**32
+        // Otherwise uint64 parsing will fail and the message is rejected
+        lastReceiptId = 4294967296;
+    }
+
+    /**
+     * Begin or increase delegation by sending an amount of ROSE to the contract.
+     *
+     * Delegation will fail if the minimum per-validator amount has not been
+     * reached, at the time of writing this is 100 ROSE.
+     *
+     * Only one delegation can occur per transaction.
+     *
+     * @param to Staking address of validator on the consensus level
+     */
+    function delegate(StakingAddress to) public payable returns (uint64) {
+        // Whatever is sent to the contract is delegated.
+        require(msg.value < type(uint128).max);
+
+        uint128 amount = uint128(msg.value);
+
+        uint64 receiptId = lastReceiptId++;
+
+        Subcall.consensusDelegate(to, amount, receiptId);
```

Comment:
This does nothing for view calls right?

Reply:
Correct, anything done in view calls is reverted as normal as with other simulated executions.